File: 1079-sum-of-root-to-leaf-binary-numbers/1079-sum-of-root-to-leaf-binary-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result = 0;
    public int sumRootToLeaf(TreeNode root) {
        return dfs(root, 0);
    }

    private static int dfs(TreeNode cur, int sum) {
        if (cur == null) return 0;
        sum = (sum << 1) + cur.val;
        if(cur.left == null && cur.right == null) return sum;
        return dfs(cur.left, sum) + dfs(cur.right, sum);
    }
}

File: 1079-sum-of-root-to-leaf-binary-numbers/1079-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:  
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        result = 0
        def dfs(cur: TreeNode, sum: int) -> int:
            if not cur: return 0
            sum += cur.val
            dfs(cur.left, sum << 1)
            dfs(cur.right, sum << 1)
            if not cur.left and not cur.right:
                nonlocal result
                result += sum
            
        dfs(root, 0)
        return result

